flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceFailover.java

```diff
@@ -333,28 +336,28 @@ private static void restartTaskManager(Runnable afterFailAction, MiniCluster min
   private static class RecordCounterToFail {
 
     private static AtomicInteger records;
-    private static CompletableFuture<Void> fail;
+    private static CountDownLatch countDownLatch;
     private static CompletableFuture<Void> continueProcessing;
 
     private static <T> DataStream<T> wrapWithFailureAfter(DataStream<T> stream, int failAfter) {
 
       records = new AtomicInteger();
-      fail = new CompletableFuture<>();
       continueProcessing = new CompletableFuture<>();
+      countDownLatch = new CountDownLatch(stream.getParallelism());
```

Comment:
curious how is the countdown latch used by the map operator. will it be serialized and a new cloned object is deserialized on the operator/subtask side? if yes, I guess there is no difference with completable future. If no, this ensures all subtasks are running and processed some records before failure is triggered.

could this be verified by break point in IDE?

Reply:
Tested with checkpoints. The static variables are "global" in the sense, that they are shared between every instance in the JVM. So the latch is used for both operator instances.

Reply:
thanks for confirming. assume you meant breakpoints (not checkpoints)

Reply:
Yes, brakepoints 😄